drivers/serial/uart_imx.c

```diff
@@ -0,0 +1,575 @@
+{
+	UART_Type *uart = UART_STRUCT(dev);
+
+	UART_SetIntCmd(uart, uartIntRxReady, true);
+}
+
+static void uart_imx_irq_rx_disable(struct device *dev)
+{
+	UART_Type *uart = UART_STRUCT(dev);
+
+	UART_SetIntCmd(uart, uartIntRxReady, false);
+}
+
+static int uart_imx_irq_rx_ready(struct device *dev)
+{
+	UART_Type *uart = UART_STRUCT(dev);
+
+	return UART_GetStatusFlag(uart,uartStatusRxReady);
+}
+
+static void uart_imx_irq_err_enable(struct device *dev)
+{
+	UART_Type *uart = UART_STRUCT(dev);
+
+	UART_SetIntCmd(uart, uartIntParityError, true);
+	UART_SetIntCmd(uart, uartIntFrameError, true);
+}
+
+static void uart_imx_irq_err_disable(struct device *dev)
+{
+	UART_Type *uart = UART_STRUCT(dev);
+
+	UART_SetIntCmd(uart, uartIntParityError, false);
+	UART_SetIntCmd(uart, uartIntFrameError, false);
+}
+
+static int uart_imx_irq_is_pending(struct device *dev)
+{
+	UART_Type *uart = UART_STRUCT(dev);
+
+	return UART_GetStatusFlag(uart,uartStatusRxReady) || UART_GetStatusFlag(uart,uartStatusTxReady);
+}
+
+static int uart_imx_irq_update(struct device *dev)
+{
+	return 1;
+}
+
+static void uart_imx_irq_callback_set(struct device *dev,
+					 uart_irq_callback_t cb)
+{
+	struct uart_imx_dev_data_t * const dev_data = DEV_DATA(dev);
+
+	dev_data->cb = cb;
+}
+
+/**
+ * @brief Interrupt service routine.
+ *
+ * This simply calls the callback function, if one exists.
+ *
+ * Note: imx UART Tx interrupts when ready to send; Rx interrupts when char
+ * received.
+ *
+ * @param arg Argument to ISR.
+ *
+ * @return N/A
+ */
+void uart_imx_isr(void *arg)
+{
+	struct device *dev = arg;
+	struct uart_imx_dev_data_t * const dev_data = DEV_DATA(dev);
+
+	if (dev_data->cb) {
+		dev_data->cb(dev);
+	}
+
+}
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+static const struct uart_driver_api uart_imx_driver_api = {
+	.poll_in =	uart_imx_poll_in,
+	.poll_out = uart_imx_poll_out,
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.fifo_fill	  = uart_imx_fifo_fill,
+	.fifo_read	  = uart_imx_fifo_read,
+	.irq_tx_enable	  = uart_imx_irq_tx_enable,
+	.irq_tx_disable   = uart_imx_irq_tx_disable,
+	.irq_tx_ready	  = uart_imx_irq_tx_ready,
+	.irq_rx_enable	  = uart_imx_irq_rx_enable,
+	.irq_rx_disable   = uart_imx_irq_rx_disable,
+	.irq_rx_ready	  = uart_imx_irq_rx_ready,
+	.irq_err_enable   = uart_imx_irq_err_enable,
+	.irq_err_disable  = uart_imx_irq_err_disable,
+	.irq_is_pending   = uart_imx_irq_is_pending,
+	.irq_update   = uart_imx_irq_update,
+	.irq_callback_set = uart_imx_irq_callback_set,
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+};
+
+
+#ifdef CONFIG_UART_IMX_UART_1
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+static void irq_config_func_1(struct device *port);
+#endif
+
+static struct uart_imx_dev_data_t uart_imx_dev_data_1 = {
+	.baud_rate = CONFIG_UART_IMX_UART_1_BAUD_RATE,
+	.modem_mode = CONFIG_UART_IMX_UART_1_MODEM_MODE,
+};
+
+static const struct uart_device_config uart_imx_dev_cfg_1 = {
+	.base = (uint8_t *)CONFIG_UART_IMX_UART_1_BASE_ADDRESS,
+	.sys_clk_freq = MHZ(100),
```

Comment:
Why is each instance set to MHZ(100), MHZ(200), ...? Are you using sys_clk_freq anywhere?